internal/cephfs/core/snapshot_metadata.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	fsAdmin "github.com/ceph/go-ceph/cephfs/admin"
+)
+
+// ErrSubVolSnapMetadataNotSupported is returned when set/get/list/remove
+// subvolume snapshot metadata options are not supported.
+var ErrSubVolSnapMetadataNotSupported = errors.New("subvolume snapshot metadata operations are not supported")
+
+func (s *snapshotClient) supportsSubVolSnapMetadata() bool {
+	return clusterAdditionalInfo[s.clusterID].subVolSnapshotMetadataState != unsupported
+}
+
+func (s *snapshotClient) isUnsupportedSubVolSnapMetadata(err error) bool {
+	var invalid fsAdmin.NotImplementedError
+	if err != nil && errors.Is(err, &invalid) {
+		// In case the error is other than invalid command return error to
+		// the caller.
+		clusterAdditionalInfo[s.clusterID].subVolSnapshotMetadataState = unsupported
+
+		return false
+	}
+	clusterAdditionalInfo[s.clusterID].subVolSnapshotMetadataState = supported
+
+	return true
+}
+
+// SetSnapshotMetadata sets custom metadata on the subvolume snapshot in a
+// volume as a key-value pair.
+func (s *snapshotClient) SetSnapshotMetadata(key, value string) error {
```

Comment:
we dont need to export this function as its not used outside of this file?

Reply:
I understand the concern. But to be practical, I wanted to align this just like SetMetadata()/GetMetadata/RemoveMetadata(). Let me know if you are not convinced.

Reply:
we(at least I) might have missed that one. if the function is not used outside of the package, it should not be exported, we can revisit the existing functions and correct them.

Reply:
Sure I can fix it.
I thought we were writing those functions assuming there will some direct consumers in the future calling them from outside this file.

Reply:
Yes but IMHO as there are no real users we shouldn't expose it, expose it when/if required always